test/runtime/kafka.go

```diff
@@ -114,32 +129,34 @@ var _ = Describe("RuntimeValidatedKafka", func() {
 		Expect(endPoints[helpers.Enabled]).To(Equal(1))
 		Expect(endPoints[helpers.Disabled]).To(Equal(2))
 
+		By("Creating kafka topics")
 		createTopic(allowedTopic)
 		createTopic(disallowTopic)
 
+		By("Listing created Kafka topics")
 		res := vm.ContainerExec("client",
 			"/opt/kafka/bin/kafka-topics.sh --list --zookeeper zook:2181")
 		res.ExpectSuccess("Cannot get kafka topics")
 
+		// Waiting for kafka broker to be up.
+		err = waitForKafkaBroker("client", produceCmd)
+		Expect(err).To(BeNil(), "Kafka broker failed to come up")
+
 		By("Allowed topic")
-		ctx, cancel := context.WithCancel(context.Background())
-		data := vm.ExecContext(ctx, consumer(allowedTopic, MaxMessages))
 
-		//TODO: wait until ready GH #3116
-		helpers.Sleep(10)
-		for i := 1; i <= MaxMessages; i++ {
+		By("Sending produce request on kafka topic `allowedTopic`")
+		for i := 1; i <= MaxMessages-1; i++ {
 			producer(allowedTopic, fmt.Sprintf("Message %d", i))
 		}
-		cancel()
 
-		err = data.WaitUntilMatch(fmt.Sprintf("Processed a total of %d messages", MaxMessages))
-		Expect(err).To(BeNil())
+		By("Sending consume request on kafka topic `allowedTopic`")
+		res = consumer(allowedTopic, MaxMessages)
+		res.ExpectSuccess("Failed to consume messages from kafka topic `allowedTopic`")
 
-		Expect(data.Output().String()).Should(ContainSubstring(
+		Expect(res.Output().String()).Should(ContainSubstring(
```

Comment:
Quick one, if you send MaxMessages -1, this assert is wrong no?

Reply:
The first message is from the test produce request in `waitForKafkaBroker` and I am consuming  `MaxMessages`, so this test is correct